OSDOCS#8861: Semver support and updating Operators

michaelryanpeter opened the pull request:
<!--- PR title format: [GH#<gh-issue-id>][BZ#<bz-issue-id>][OCPBUGS#<jira-issue-id>][OSDOCS#<jira-issue-id>]: <short-description-of-the-pr> --->

<!--- If your changes apply to the latest release and/or in-development version of OpenShift, open your PR against the `main` branch.

* For more details about the information requested in this template, see:
  https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/create_or_edit_content.adoc#submit-PR --->

Version(s): 4.15
<!--- Specify the version or versions of OpenShift your PR applies to. -->

Issue: [OSDOCS-8861](https://issues.redhat.com//browse/OSDOCS-8861), [OSDOCS-8862](https://issues.redhat.com//browse/OSDOCS-8862), & [OSDOCS-8866](https://issues.redhat.com//browse/OSDOCS-8866)
<!--- Add a link to the Bugzilla, Jira, or GitHub issue, if applicable. --->

Link to docs preview:
* [Installing an Operator (small tweaks)](https://71258--ocpdocs-pr.netlify.app/openshift-enterprise/latest/operators/olm_v1/olmv1-installing-an-operator-from-a-catalog#olmv1-installing-an-operator_olmv1-installing-operator)
* [Updating Operators](https://71258--ocpdocs-pr.netlify.app/openshift-enterprise/latest/operators/olm_v1/olmv1-installing-an-operator-from-a-catalog#olmv1-updating-an-operator_olmv1-installing-operator)
* [Semver support](https://71258--ocpdocs-pr.netlify.app/openshift-enterprise/latest/operators/olm_v1/olmv1-installing-an-operator-from-a-catalog#olmv1-semver-support_olmv1-installing-operator)
* [Example CRs](https://71258--ocpdocs-pr.netlify.app/openshift-enterprise/latest/operators/olm_v1/olmv1-installing-an-operator-from-a-catalog#olmv1-about-operator-updates_olmv1-installing-operator)

<!--- Add direct link(s) to the exact page(s) with updated content from the preview build. --->

QE review:
- [x] QE has approved this change.
<!--- QE approval is required to merge a PR except for changes that do not impact the meaning of the docs. --->

Additional information:
<!--- Optional: Include additional context or expand the description here.--->

<!--- After you open your PR, ask for review from the OpenShift docs team:
  For community authors: Tag @openshift/team-documentation in a GitHub comment.--->

michaelryanpeter commented:
@jianzhangbjz or @kuiwang02: PTAL

kevinrizza commented on modules/olmv1-about-target-versions.adoc:
```diff
@@ -3,106 +3,70 @@
+If you specify the Operator or extension's target version in the CR, {olmv1} installs the specified version. When the target version is specified in the CR, {olmv1} does not change the target version when updates are published to the catalog.
 
-If you want to update the version of the Operator that is installed on the cluster, you must manually update the Operator's CR. Specifying a Operator's target version pins the Operator's version to the specified release.
+If you want to update the version of the Operator that is installed on the cluster, you must manually edit the Operator's CR. Specifying a Operator's target version pins the Operator's version to the specified release.
```
`Specifying an Operator's` instead of `Specifying a Operator's`?

kevinrizza commented on modules/olmv1-about-target-versions.adoc:
```diff
@@ -3,106 +3,70 @@
-====
-In previous versions of OLM, Operator authors could define upgrade edges to prevent you from updating to unsupported versions. In its current state of development, {olmv1} does not enforce upgrade edge definitions. You can specify any version of an Operator, and {olmv1} attempts to apply the update.
-====
+If you want to define update edges for an Operator or extension that follows that semantic versioning (semver) standard, you can specify a version range using a comparison string.
```
You're not really `defining update edges` by setting a version range, you're specifying a version range that will be resolved to the highest valid value by the operator controller.

kevinrizza commented on modules/olmv1-semver-support.adoc:
```diff
@@ -0,0 +1,32 @@
+
+Support for link:https://semver.org/[semantic versioning (semver)] is enabled in {olmv1} by default. Operator and extension authors can use the semver standard to define compatible updates.
+
+According to the semver standard, releases within a major version number, such as `>=1.0.0, <2.0.0`, must be compatible. As a result, Operator or extension authors do not need to define update edges explicitly. {olmv1-first} can use an Operator or extension's version number to determine if an update can be resolved successfully.
```
I might be less explicit about this section? I think we haven't landed on a design that definitively sets this as a standard, we just haven't implemented constraint resolution based on the existing upgrade edge spec yet. @joelanford ?

kevinrizza commented on modules/olmv1-semver-support.adoc:
```diff
@@ -0,0 +1,32 @@
+
+According to the semver standard, releases within a major version number, such as `>=1.0.0, <2.0.0`, must be compatible. As a result, Operator or extension authors do not need to define update edges explicitly. {olmv1-first} can use an Operator or extension's version number to determine if an update can be resolved successfully.
+
+Cluster administrators can easily define the update edges for Operators and extensions that follow the semver standard by using comparison strings to define to specify a desired version range.
```
Same comment as above. Users aren't defining the upgrade edges, they're just specifying a range for versions that may or may not be valid to upgrade to based on the controller's resolution.

kevinrizza reviewed (commented):
This mostly looks good! I just had a couple of comments, mostly around the semantics of how we express upgrades

michaelryanpeter commented:
Thanks, @kevinrizza. I addressed your feedback in https://github.com/openshift/openshift-docs/pull/71258/commits/5ee1deab35e64b48a87b3d33e9919e03e5fe9077. Let me know if these changes address your concerns.

kevinrizza commented:
lgtm

jianzhangbjz commented:
/lgtm

michaelryanpeter commented:
/label peer-review-needed

m1kola commented on modules/olmv1-updating-an-operator.adoc:
```diff
@@ -7,71 +7,186 @@
 
 * You have a catalog installed.
-* You have an Operator installed.
+* You have downloaded a local copy of the catalog file.
```
Do we describe how to do it somewhere?

michaelryanpeter replied:
We describe this earlier in the doc, here: https://71258--ocpdocs-pr.netlify.app/openshift-enterprise/latest/operators/olm_v1/olmv1-installing-an-operator-from-a-catalog#olmv1-finding-operators-to-install_olmv1-installing-operator

m1kola commented on modules/olmv1-about-target-versions.adoc:
```diff
@@ -3,106 +3,70 @@
 [source,yaml]
 ----
 apiVersion: operators.operatorframework.io/v1alpha1
 kind: Operator
```
I believe `Operator` was renamed to `ClusterExtension` since we introduced semver support.

Should we update docs (separate PR is ok for this) to reflect the change?

michaelryanpeter replied:
The change to `ClusterExtension` did not make the cut off for 4.15. I think the change went in right after we branched for 4.16. Bryce looked it up to confirm.

m1kola reviewed (approved):
Looks good to me. Just a couple of questions which can be addressed in followup PRs.

aldiazRH commented:
/remove-label peer-review-needed
/label peer-review-in-progress

aldiazRH commented on modules/olmv1-about-target-versions.adoc:
```diff
@@ -3,106 +3,70 @@
-====
-In previous versions of OLM, Operator authors could define upgrade edges to prevent you from updating to unsupported versions. In its current state of development, {olmv1} does not enforce upgrade edge definitions. You can specify any version of an Operator, and {olmv1} attempts to apply the update.
-====
+If you want to define a range of acceptable versions for an Operator or extension, you can specify a version range by using a comparison string. When you specify a version range, {olmv1} installs a highest version of an Operator or extension that can be resolved by the Operator Controller.
```
```suggestion
If you want to define a range of acceptable versions for an Operator or extension, you can specify a version range by using a comparison string. When you specify a version range, {olmv1} installs the latest version of an Operator or extension that can be resolved by the Operator Controller.
```
Would something like this work better?

michaelryanpeter replied:
Great suggestion. Thanks, @aldiazRH!

aldiazRH reviewed (commented):
One tiny nit, but otherwise LGTM!

aldiazRH commented:
/label peer-review-done
/remove-label peer-review-in-progress

michaelryanpeter commented:
/cherrypick enterprise-4.15
